src/vm/scanning.rs

```diff
@@ -11,6 +10,31 @@ pub trait EdgeVisitor {
     // TODO: Add visit_soft_edge, visit_weak_edge, ... here.
 }
 
+/// Root-scanning methods use this trait to spawn work packets for processing roots.
+pub trait RootsWorkFactory: Send {
+    /// Create work packets to handle the roots represented as edges.
+    ///
+    /// The work packet may update the edge.
+    ///
+    /// Arguments:
+    /// * `edges`: A vector of edges.
+    fn create_process_edge_roots_work(&self, edges: Vec<Address>);
+
+    /// Create work packets to handle edges.
+    ///
+    /// The work packet cannot update the roots.  This is a good chance to pin the objects.
+    ///
+    /// This method is useful for conservative stack scanning, or VMs that cannot update some
+    /// of the root edges.
+    ///
+    /// Arguments:
+    /// * `nodes`: A vector of object references pointed by root edges.
+    fn create_process_node_roots_work(&self, nodes: Vec<ObjectReference>);
+
+    /// Create a copy of the factory itself.
+    fn fork(&self) -> Box<dyn RootsWorkFactory>;
```

Comment:
Why not make `RootsWorkFactory: Sync`, and use `Arc<dyn RootsWorkFactory>` to pass the factory to the binding? It will be more clear that the factory can be shared among threads.

Reply:
Since `ProcessEdgesWorkRootsWorkFactory` has only one field `mmtk: &'static MMTK<E::VM>`, we can further make `RootsWorkFactory: Copy` to avoid heap allocation.

Reply:
Yes.  I thought cloning an `Arc` would be more expensive than cloning a `Box` because of the underlying atomic memory operation.  But after a second thought, I think that cost should be trivial, and the `RootsWorkFactory` should be freely sharable among threads because both of the required `create_process_{node,edge}_roots_work` methods take an immutable `&self` reference.

I'll try to change it to Arc.

Reply:
`Arc` would be slightly different than this `fork()` model:
1. With `Arc`, the factory probably won't have any mutable state (unless lock is used).
2. If the factory has any state, `fork()` may choose to create a fresh instance or inherit the state. But with `Arc`, as they point to the same instance, the state is shared.

I am not sure which is more suitable when we have other  `RootsWorkFactory` impls.